File: Front/src/types/quota.ts
// 🎯 Quota Management TypeScript Types
// Complete type definitions for department quota management

// =============================================================================
// ENUMS - MATCHING BACKEND EXACTLY
// =============================================================================

/**
 * Types of quotas that can be set
 * 
 * Learning: These match the backend QuotaType enum exactly.
 * - COST: Dollar amount limits (most common for budget control)
 * - TOKENS: Token count limits (for usage control)
 * - REQUESTS: Request count limits (for rate limiting)
 */
export type QuotaType = 'cost' | 'tokens' | 'requests';

/**
 * How often quotas reset
 * 
 * Learning: Different departments need different reset cycles.
 * - DAILY: For tight control or testing
 * - WEEKLY: For project-based teams
 * - MONTHLY: Most common for budget allocation
 * - YEARLY: For annual planning
 */
export type QuotaPeriod = 'daily' | 'weekly' | 'monthly' | 'yearly';

/**
 * Current status of a quota
 * 
 * Learning: Status helps admins understand quota health at a glance.
 */
export type QuotaStatus = 'active' | 'inactive' | 'suspended' | 'exceeded';

// =============================================================================
// CORE QUOTA INTERFACES
// =============================================================================

/**
 * Complete quota information returned by the API
 * 
 * Learning: This matches the backend QuotaResponse schema exactly.
 * All the calculated fields (usage_percentage, is_exceeded, etc.) 
 * are computed by the backend for consistency.
 */
export interface QuotaResponse {
  // Basic quota information
  id: number;
  name: string;
  description?: string;
  
  // Department and LLM configuration relationships
  department_id: number;
  department_name?: string;
  llm_config_id?: number; // null means "all providers"
  llm_config_name?: string;
  
  // Quota configuration
  quota_type: string; // 'cost', 'tokens', 'requests'
  quota_period: string; // 'daily', 'weekly', 'monthly', 'yearly'
  limit_value: number; // The maximum allowed amount
  
  // Current usage and status
  current_usage: number; // How much has been used this period
  remaining_quota: number; // How much is left
  usage_percentage: number; // Percentage used (0-100+)
  
  // Status flags
  status: string; // 'active', 'inactive', 'suspended', 'exceeded'
  is_enforced: boolean; // Whether this quota blocks requests when exceeded
  is_exceeded: boolean; // Whether current usage > limit
  is_near_limit: boolean; // Whether usage is close to limit (usually 80%+)
  
  // Time information
  period_start?: string; // When current period started (ISO date)
  period_end?: string; // When current period ends (ISO date)
  next_reset_at?: string; // When quota will next reset (ISO date)
  
  // Audit information
  created_by: string; // Email of admin who created this quota
  created_at: string; // Creation timestamp (ISO date)
  updated_at: string; // Last update timestamp (ISO date)
}

/**
 * Data needed to create a new quota
 * 
 * Learning: This matches the backend QuotaCreateRequest schema.
 * Notice how we use the raw enum types for the form data.
 */
export interface QuotaCreateRequest {
  department_id: number; // Required: which department does this apply to
  llm_config_id?: number; // Optional: specific LLM provider (null = all providers)
  quota_type: QuotaType; // Required: cost, tokens, or requests
  quota_period: QuotaPeriod; // Required: how often it resets
  limit_value: number; // Required: the actual limit amount
  name: string; // Required: human-readable name
  description?: string; // Optional: more details about the quota
  is_enforced?: boolean; // Optional: whether to block when exceeded (default: true)
}

/**
 * Data that can be updated for an existing quota
 * 
 * Learning: All fields optional for partial updates.
 * This matches the backend QuotaUpdateRequest schema.
 */
export interface QuotaUpdateRequest {
  limit_value?: number; // Update the limit amount
  name?: string; // Update the name
  description?: string; // Update the description
  is_enforced?: boolean; // Update enforcement setting
  status?: QuotaStatus; // Update the status
  quota_period?: QuotaPeriod; // Update reset period
}

// =============================================================================
// LIST AND PAGINATION TYPES
// =============================================================================

/**
 * Filters for searching and filtering quotas
 * 
 * Learning: Comprehensive filtering helps admins find specific quotas quickly.
 * This matches the backend query parameters.
 */
export interface QuotaSearchFilters {
  // Basic filters
  department_id?: number;
  llm_config_id?: number;
  quota_type?: QuotaType;
  quota_period?: QuotaPeriod;
  status?: QuotaStatus;
  is_enforced?: boolean;
  is_exceeded?: boolean;
  
  // Text search
  search?: string; // Search in names and descriptions
  
  // Pagination
  page?: number; // Page number (1-based)
  page_size?: number; // Items per page
  
  // Sorting
  sort_by?: string; // Field to sort by
  sort_order?: 'asc' | 'desc'; // Sort direction
}

/**
 * Paginated list of quotas with metadata
 * 
 * Learning: This includes both the quota data and pagination info
 * needed to build a complete UI with page navigation.
 */
export interface QuotaListResponse {
  quotas: QuotaResponse[]; // Array of quota objects
  
  // Pagination metadata
  total_count: number; // Total quotas matching filters
  page: number; // Current page number
  page_size: number; // Items per page
  total_pages: number; // Total number of pages
  has_next: boolean; // Whether there are more pages
  has_previous: boolean; // Whether there are previous pages
  
  // Summary statistics for the current filter results
  summary: {
    total_quotas: number;
    active_quotas: number;
    exceeded_quotas: number;
    near_limit_quotas: number;
    total_monthly_cost_limit: number;
    total_monthly_cost_used: number;
  };
}

// =============================================================================
// DEPARTMENT QUOTA STATUS TYPES
// =============================================================================

/**
 * Comprehensive status for all quotas in a department
 * 
 * Learning: This provides a department-focused view of quota health.
 * Useful for department managers and budget oversight.
 */
export interface DepartmentQuotaStatusResponse {
  department_id: number;
  department_name?: string;
  last_updated: string; // ISO timestamp
  
  overall_status: string; // 'healthy', 'warning', 'exceeded', 'no_quotas'
  
  // Summary counts
  total_quotas: number;
  active_quotas: number;
  exceeded_quotas: number;
  near_limit_quotas: number;
  suspended_quotas: number;
  inactive_quotas: number;
  
  // Aggregated statistics by quota type
  quotas_by_type: Record<string, {
    count: number;
    total_limit: number;
    total_usage: number;
  }>;
  
  // Individual quota details
  quotas: QuotaResponse[];
}

// =============================================================================
// QUOTA OPERATIONS TYPES
// =============================================================================

/**
 * Response from resetting a quota
 * 
 * Learning: Reset operations provide detailed feedback about what happened.
 */
export interface QuotaResetResponse {
  success: boolean;
  message: string;
  quota_id: number;
  quota_name: string;
  reset_at: string; // ISO timestamp
  previous_usage: number;
  new_usage: number;
}

/**
 * Response from bulk operations on multiple quotas
 * 
 * Learning: Bulk operations need detailed reporting so admins know
 * exactly what succeeded and what failed.
 */
export interface BulkQuotaOperationResponse {
  success: boolean; // True if ALL operations succeeded
  message: string;
  total_requested: number;
  successful_operations: number;
  failed_operations: number;
  results: Array<{
    quota_id: number;
    quota_name?: string;
    success: boolean;
    message: string;
  }>;
}

// =============================================================================
// ANALYTICS TYPES
// =============================================================================

/**
 * Quota system analytics summary
 * 
 * Learning: High-level metrics help admins understand quota system health.
 */
export interface QuotaAnalyticsSummary {
  overview: {
    total_quotas: number;
    active_quotas: number;
    exceeded_quotas: number;
    health_percentage: number; // Percentage of quotas that are healthy
  };
  
  by_type: Record<string, {
    count: number;
    total_limit: number;
    total_usage: number;
    utilization_percentage: number;
  }>;
  
  by_period: Record<string, number>; // Count of quotas by reset period
  
  top_departments: Array<{
    department_name: string;
    quota_count: number;
    total_usage: number;
  }>;
  
  last_updated: string; // ISO timestamp
}

// =============================================================================
// UI HELPER TYPES
// =============================================================================

/**
 * Form state for quota creation/editing
 * 
 * Learning: Separating form state helps manage complex forms with validation.
 */
export interface QuotaFormState {
  data: QuotaCreateRequest | QuotaUpdateRequest;
  errors: QuotaFormErrors;
  isSubmitting: boolean;
  isDirty: boolean;
}

/**
 * Form validation errors
 * 
 * Learning: Field-specific errors provide better user experience.
 */
export interface QuotaFormErrors {
  department_id?: string;
  llm_config_id?: string;
  quota_type?: string;
  quota_period?: string;
  limit_value?: string;
  name?: string;
  description?: string;
  is_enforced?: string;
  submit?: string; // General form-level errors
}

/**
 * Options for quota type dropdowns
 * 
 * Learning: UI components need human-readable labels and descriptions.
 */
export interface QuotaTypeOption {
  value: QuotaType;
  label: string;
  description: string;
  icon?: string;
}

export interface QuotaPeriodOption {
  value: QuotaPeriod;
  label: string;
  description: string;
  icon?: string;
}

// =============================================================================
// DEPARTMENT AND LLM CONFIG REFERENCE TYPES
// =============================================================================

/**
 * Simplified department info for quota forms
 * 
 * Learning: We need basic department info for dropdown selections.
 * This avoids importing the full admin types in quota components.
 */
export interface DepartmentOption {
  id: number;
  name: string;
  code?: string;
  is_active: boolean;
}

/**
 * Simplified LLM config info for quota forms
 * 
 * Learning: Similar to department - just what we need for form dropdowns.
 */
export interface LLMConfigOption {
  id: number;
  name: string;
  provider: string;
  is_active: boolean;
}

// =============================================================================
// ERROR HANDLING TYPES
// =============================================================================

/**
 * Custom error class for quota API operations
 * 
 * Learning: Custom errors provide better debugging and user feedback.
 */
export class QuotaApiError extends Error {
  constructor(
    message: string,
    public status: number = 500,
    public details?: any
  ) {
    super(message);
    this.name = 'QuotaApiError';
  }
}

// =============================================================================
// CONSTANTS FOR UI
// =============================================================================

/**
 * Predefined quota type options for UI components
 * 
 * Learning: Centralizing UI constants makes the interface consistent.
 */
export const QUOTA_TYPE_OPTIONS: QuotaTypeOption[] = [
  {
    value: 'cost',
    label: 'Cost (USD)',
    description: 'Limit spending in dollars',
    icon: '💰'
  },
  {
    value: 'tokens',
    label: 'Tokens',
    description: 'Limit AI processing tokens',
    icon: '🎯'
  },
  {
    value: 'requests',
    label: 'Requests',
    description: 'Limit number of API calls',
    icon: '📞'
  }
];

/**
 * Predefined quota period options for UI components
 */
export const QUOTA_PERIOD_OPTIONS: QuotaPeriodOption[] = [
  {
    value: 'daily',
    label: 'Daily',
    description: 'Resets every day at midnight',
    icon: '📅'
  },
  {
    value: 'weekly',
    label: 'Weekly',
    description: 'Resets every Monday',
    icon: '📆'
  },
  {
    value: 'monthly',
    label: 'Monthly',
    description: 'Resets on the 1st of each month',
    icon: '🗓️'
  },
  {
    value: 'yearly',
    label: 'Yearly',
    description: 'Resets on January 1st',
    icon: '📋'
  }
];

// =============================================================================
// UTILITY TYPE GUARDS
// =============================================================================

/**
 * Type guard to check if a quota is exceeded
 * 
 * Learning: Type guards provide runtime type safety and better code clarity.
 */
export function isQuotaExceeded(quota: QuotaResponse): boolean {
  return quota.is_exceeded || quota.current_usage >= quota.limit_value;
}

/**
 * Type guard to check if a quota is near its limit
 */
export function isQuotaNearLimit(quota: QuotaResponse): boolean {
  return quota.is_near_limit || quota.usage_percentage >= 80;
}

/**
 * Type guard to check if a quota is healthy
 */
export function isQuotaHealthy(quota: QuotaResponse): boolean {
  return quota.status === 'active' && !isQuotaExceeded(quota) && !isQuotaNearLimit(quota);
}

// =============================================================================
// EXPORT ALL TYPES
// =============================================================================

export type {
  // Core types
  QuotaResponse,
  QuotaCreateRequest,
  QuotaUpdateRequest,
  QuotaSearchFilters,
  QuotaListResponse,
  
  // Status and analytics
  DepartmentQuotaStatusResponse,
  QuotaAnalyticsSummary,
  
  // Operations
  QuotaResetResponse,
  BulkQuotaOperationResponse,
  
  // UI helpers
  QuotaFormState,
  QuotaFormErrors,
  DepartmentOption,
  LLMConfigOption,
  
  // Utility types
  QuotaType,
  QuotaPeriod,
  QuotaStatus
};

File: Back/app/schemas/quota.py
# AI Dock Quota API Schemas
# Pydantic models for quota-related API requests and responses

from pydantic import BaseModel, Field, field_validator
from typing import List, Optional, Dict, Any
from decimal import Decimal
from datetime import datetime
from enum import Enum

from ..models.quota import QuotaType, QuotaPeriod, QuotaStatus

# =============================================================================
# REQUEST SCHEMAS - WHAT CLIENTS SEND TO THE API
# =============================================================================

class QuotaCreateRequest(BaseModel):
    """
    Schema for creating a new department quota.
    
    This defines what data is required when an admin wants to create
    a new quota through the API.
    """
    
    department_id: int = Field(
        ..., 
        description="ID of the department this quota applies to",
        gt=0
    )
    
    llm_config_id: Optional[int] = Field(
        None,
        description="ID of LLM configuration (null for all providers)",
        gt=0
    )
    
    quota_type: QuotaType = Field(
        ...,
        description="Type of quota: cost, tokens, or requests"
    )
    
    quota_period: QuotaPeriod = Field(
        ...,
        description="Reset period: daily, weekly, monthly, or yearly"
    )
    
    limit_value: Decimal = Field(
        ...,
        description="The quota limit amount",
        gt=0,
        max_digits=15,
        decimal_places=4
    )
    
    name: str = Field(
        ...,
        description="Human-readable name for this quota",
        min_length=1,
        max_length=200
    )
    
    description: Optional[str] = Field(
        None,
        description="Optional description of the quota purpose",
        max_length=500
    )
    
    is_enforced: bool = Field(
        True,
        description="Whether to enforce this quota (block requests when exceeded)"
    )
    
    @field_validator('name')
    @classmethod
    def name_must_not_be_empty(cls, v):
        """Ensure name is not just whitespace"""
        if not v or not v.strip():
            raise ValueError('Name cannot be empty or whitespace')
        return v.strip()
    
    class Config:
        json_schema_extra = {
            "example": {
                "department_id": 1,
                "llm_config_id": 1,
                "quota_type": "cost",
                "quota_period": "monthly",
                "limit_value": 1000.00,
                "name": "Engineering OpenAI Monthly Budget",
                "description": "Monthly spending limit for Engineering on OpenAI GPT-4",
                "is_enforced": True
            }
        }

class QuotaUpdateRequest(BaseModel):
    """
    Schema for updating an existing quota.
    
    All fields are optional so admins can update just what they need.
    """
    
    limit_value: Optional[Decimal] = Field(
        None,
        description="New quota limit amount",
        gt=0,
        max_digits=15,
        decimal_places=4
    )
    
    name: Optional[str] = Field(
        None,
        description="New quota name",
        min_length=1,
        max_length=200
    )
    
    description: Optional[str] = Field(
        None,
        description="New quota description",
        max_length=500
    )
    
    is_enforced: Optional[bool] = Field(
        None,
        description="Whether to enforce this quota"
    )
    
    status: Optional[QuotaStatus] = Field(
        None,
        description="New quota status"
    )
    
    quota_period: Optional[QuotaPeriod] = Field(
        None,
        description="New reset period"
    )
    
    @field_validator('name')
    @classmethod
    def name_must_not_be_empty(cls, v):
        """Ensure name is not just whitespace if provided"""
        if v is not None and (not v or not v.strip()):
            raise ValueError('Name cannot be empty or whitespace')
        return v.strip() if v else v
    
    class Config:
        json_schema_extra = {
            "example": {
                "limit_value": 1500.00,
                "name": "Engineering OpenAI Monthly Budget (Updated)",
                "is_enforced": True
            }
        }

class QuotaFilterRequest(BaseModel):
    """
    Schema for filtering quotas in list requests.
    
    This allows admins to search and filter quotas by various criteria.
    """
    
    department_id: Optional[int] = Field(
        None,
        description="Filter by department ID",
        gt=0
    )
    
    llm_config_id: Optional[int] = Field(
        None,
        description="Filter by LLM configuration ID",
        gt=0
    )
    
    quota_type: Optional[QuotaType] = Field(
        None,
        description="Filter by quota type"
    )
    
    quota_period: Optional[QuotaPeriod] = Field(
        None,
        description="Filter by reset period"
    )
    
    status: Optional[QuotaStatus] = Field(
        None,
        description="Filter by quota status"
    )
    
    is_enforced: Optional[bool] = Field(
        None,
        description="Filter by enforcement status"
    )
    
    is_exceeded: Optional[bool] = Field(
        None,
        description="Filter by whether quota is exceeded"
    )
    
    search: Optional[str] = Field(
        None,
        description="Search in quota names and descriptions",
        max_length=100
    )
    
    # Pagination
    page: int = Field(
        1,
        description="Page number (1-based)",
        ge=1
    )
    
    page_size: int = Field(
        20,
        description="Number of items per page",
        ge=1,
        le=100
    )
    
    # Sorting
    sort_by: Optional[str] = Field(
        "name",
        description="Field to sort by: name, created_at, limit_value, usage_percentage"
    )
    
    sort_order: Optional[str] = Field(
        "asc",
        description="Sort order: asc or desc"
    )
    
    @field_validator('sort_by')
    @classmethod
    def validate_sort_field(cls, v):
        """Ensure sort field is valid"""
        allowed_fields = {'name', 'created_at', 'limit_value', 'usage_percentage', 'department_name'}
        if v and v not in allowed_fields:
            raise ValueError(f'Sort field must be one of: {", ".join(allowed_fields)}')
        return v
    
    @field_validator('sort_order')
    @classmethod
    def validate_sort_order(cls, v):
        """Ensure sort order is valid"""
        if v and v.lower() not in ['asc', 'desc']:
            raise ValueError('Sort order must be "asc" or "desc"')
        return v.lower() if v else v

# =============================================================================
# RESPONSE SCHEMAS - WHAT THE API SENDS BACK TO CLIENTS
# =============================================================================

class QuotaResponse(BaseModel):
    """
    Schema for quota information in API responses.
    
    This is the standardized format for returning quota data.
    """
    
    id: int
    department_id: int
    department_name: Optional[str] = None
    llm_config_id: Optional[int] = None
    llm_config_name: Optional[str] = None
    
    quota_type: str
    quota_period: str
    
    limit_value: float
    current_usage: float
    remaining_quota: float
    usage_percentage: float
    
    status: str
    is_enforced: bool
    is_exceeded: bool
    is_near_limit: bool
    
    name: str
    description: Optional[str] = None
    
    period_start: Optional[datetime] = None
    period_end: Optional[datetime] = None
    next_reset_at: Optional[datetime] = None
    
    created_by: str
    created_at: datetime
    updated_at: datetime
    
    class Config:
        from_attributes = True
        json_schema_extra = {
            "example": {
                "id": 1,
                "department_id": 1,
                "department_name": "Engineering",
                "llm_config_id": 1,
                "llm_config_name": "OpenAI GPT-4",
                "quota_type": "cost",
                "quota_period": "monthly",
                "limit_value": 1000.00,
                "current_usage": 750.50,
                "remaining_quota": 249.50,
                "usage_percentage": 75.05,
                "status": "active",
                "is_enforced": True,
                "is_exceeded": False,
                "is_near_limit": True,
                "name": "Engineering OpenAI Monthly Budget",
                "description": "Monthly spending limit for Engineering on OpenAI GPT-4",
                "period_start": "2024-06-01T00:00:00Z",
                "period_end": "2024-07-01T00:00:00Z",
                "next_reset_at": "2024-07-01T00:00:00Z",
                "created_by": "admin@company.com",
                "created_at": "2024-05-15T10:30:00Z",
                "updated_at": "2024-06-05T14:20:00Z"
            }
        }

class QuotaListResponse(BaseModel):
    """
    Schema for paginated list of quotas.
    
    This provides both the quota data and pagination metadata.
    """
    
    quotas: List[QuotaResponse]
    
    # Pagination metadata
    total_count: int = Field(description="Total number of quotas matching filters")
    page: int = Field(description="Current page number")
    page_size: int = Field(description="Number of items per page")
    total_pages: int = Field(description="Total number of pages")
    has_next: bool = Field(description="Whether there are more pages")
    has_previous: bool = Field(description="Whether there are previous pages")
    
    # Summary statistics
    summary: Dict[str, Any] = Field(description="Summary statistics about the quotas")
    
    class Config:
        json_schema_extra = {
            "example": {
                "quotas": [
                    # QuotaResponse examples would go here
                ],
                "total_count": 45,
                "page": 1,
                "page_size": 20,
                "total_pages": 3,
                "has_next": True,
                "has_previous": False,
                "summary": {
                    "total_quotas": 45,
                    "active_quotas": 42,
                    "exceeded_quotas": 3,
                    "near_limit_quotas": 8,
                    "total_monthly_cost_limit": 25000.00,
                    "total_monthly_cost_used": 18750.50
                }
            }
        }

class DepartmentQuotaStatusResponse(BaseModel):
    """
    Schema for comprehensive department quota status.
    
    This provides a complete overview of all quotas for a department.
    """
    
    department_id: int
    department_name: Optional[str] = None
    last_updated: datetime
    
    overall_status: str = Field(description="Overall status: healthy, warning, exceeded, no_quotas")
    
    # Quota counts
    total_quotas: int
    active_quotas: int
    exceeded_quotas: int
    near_limit_quotas: int
    suspended_quotas: int
    inactive_quotas: int
    
    # Aggregated data by quota type
    quotas_by_type: Dict[str, Dict[str, float]] = Field(
        description="Quota statistics grouped by type (cost, tokens, requests)"
    )
    
    # Individual quota details
    quotas: List[QuotaResponse]
    
    class Config:
        json_schema_extra = {
            "example": {
                "department_id": 1,
                "department_name": "Engineering",
                "last_updated": "2024-06-06T15:30:00Z",
                "overall_status": "warning",
                "total_quotas": 5,
                "active_quotas": 5,
                "exceeded_quotas": 0,
                "near_limit_quotas": 2,
                "suspended_quotas": 0,
                "inactive_quotas": 0,
                "quotas_by_type": {
                    "cost": {
                        "count": 3,
                        "total_limit": 5000.00,
                        "total_usage": 3750.25
                    },
                    "tokens": {
                        "count": 1,
                        "total_limit": 100000,
                        "total_usage": 75000
                    },
                    "requests": {
                        "count": 1,
                        "total_limit": 1000,
                        "total_usage": 850
                    }
                },
                "quotas": []
            }
        }

class QuotaResetResponse(BaseModel):
    """Schema for quota reset operation response"""
    
    success: bool
    message: str
    quota_id: int
    quota_name: str
    reset_at: datetime
    previous_usage: float
    new_usage: float = 0.0
    
    class Config:
        json_schema_extra = {
            "example": {
                "success": True,
                "message": "Quota reset successfully",
                "quota_id": 1,
                "quota_name": "Engineering OpenAI Monthly Budget",
                "reset_at": "2024-06-06T15:45:00Z",
                "previous_usage": 750.50,
                "new_usage": 0.0
            }
        }

class BulkQuotaOperationResponse(BaseModel):
    """Schema for bulk operations on multiple quotas"""
    
    success: bool
    message: str
    total_requested: int
    successful_operations: int
    failed_operations: int
    results: List[Dict[str, Any]]
    
    class Config:
        json_schema_extra = {
            "example": {
                "success": True,
                "message": "Bulk operation completed",
                "total_requested": 5,
                "successful_operations": 4,
                "failed_operations": 1,
                "results": [
                    {"quota_id": 1, "success": True, "message": "Reset successful"},
                    {"quota_id": 2, "success": True, "message": "Reset successful"},
                    {"quota_id": 3, "success": False, "message": "Quota not found"},
                    {"quota_id": 4, "success": True, "message": "Reset successful"},
                    {"quota_id": 5, "success": True, "message": "Reset successful"}
                ]
            }
        }

# =============================================================================
# ERROR RESPONSE SCHEMAS
# =============================================================================

class QuotaErrorResponse(BaseModel):
    """Schema for quota-related error responses"""
    
    error: str = Field(description="Error type")
    message: str = Field(description="Human-readable error message")
    details: Optional[Dict[str, Any]] = Field(None, description="Additional error details")
    timestamp: datetime = Field(description="When the error occurred")
    
    class Config:
        json_schema_extra = {
            "example": {
                "error": "quota_not_found",
                "message": "Quota with ID 123 not found",
                "details": {
                    "requested_id": 123,
                    "available_quotas": [1, 2, 3, 4, 5]
                },
                "timestamp": "2024-06-06T15:30:00Z"
            }
        }

class ValidationErrorResponse(BaseModel):
    """Schema for validation error responses"""
    
    error: str = "validation_error"
    message: str = Field(description="Validation error message")
    field_errors: Dict[str, List[str]] = Field(description="Field-specific validation errors")
    timestamp: datetime = Field(description="When the error occurred")
    
    class Config:
        json_schema_extra = {
            "example": {
                "error": "validation_error",
                "message": "Request validation failed",
                "field_errors": {
                    "limit_value": ["must be greater than 0"],
                    "name": ["field required"]
                },
                "timestamp": "2024-06-06T15:30:00Z"
            }
        }

# =============================================================================
# UTILITY FUNCTIONS FOR SCHEMA CONVERSION
# =============================================================================

def convert_quota_to_response(quota, department_name: str = None, llm_config_name: str = None) -> QuotaResponse:
    """
    Convert a DepartmentQuota model instance to a QuotaResponse schema.
    
    Args:
        quota: DepartmentQuota model instance
        department_name: Optional department name
        llm_config_name: Optional LLM configuration name
        
    Returns:
        QuotaResponse with all computed fields
    """
    return QuotaResponse(
        id=quota.id,
        department_id=quota.department_id,
        department_name=department_name or (quota.department.name if hasattr(quota, 'department') and quota.department else None),
        llm_config_id=quota.llm_config_id,
        llm_config_name=llm_config_name or (quota.llm_config.name if hasattr(quota, 'llm_config') and quota.llm_config else "All Providers"),
        quota_type=quota.quota_type.value,
        quota_period=quota.quota_period.value,
        limit_value=float(quota.limit_value),
        current_usage=float(quota.current_usage),
        remaining_quota=float(quota.get_remaining_quota()),
        usage_percentage=quota.get_usage_percentage(),
        status=quota.status.value,
        is_enforced=quota.is_enforced,
        is_exceeded=quota.is_exceeded(),
        is_near_limit=quota.is_near_limit(),
        name=quota.name,
        description=quota.description,
        period_start=quota.period_start,
        period_end=quota.period_end,
        next_reset_at=quota.next_reset_at,
        created_by=quota.created_by,
        created_at=quota.created_at,
        updated_at=quota.updated_at
    )
